### PR TITLE
ci: move from drone ci to github actions

### PR DESCRIPTION
**Pull request by starbops**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**

We're going to transfer to GitHub Actions from Drone CI.

**Solution:**

Migrate pipelines originally defined in `.drone.yml` to GitHub Actions workflows.

Besides the CI migration, the golangci-lint version is bumped and added a timeout flag for execution.

**Related Issue:**

harvester/harvester#5327

**Test plan:**

### CODE REVIEW

File: scripts/build [hunk 1]
```diff
@@ -12,6 +12,8 @@ mkdir -p bin
 LINKFLAGS="-X main.AppVersion=$VERSION
            -X main.GitCommit=$COMMIT"
 
-CGO_ENABLED=0 GOARCH="$ARCH" go build -ldflags "-s -w $LINKFLAGS" -o bin/vm-dhcp-controller ./cmd/controller
-CGO_ENABLED=0 GOARCH="$ARCH" go build -ldflags "-s -w $LINKFLAGS" -o bin/vm-dhcp-agent ./cmd/agent
-CGO_ENABLED=0 GOARCH="$ARCH" go build -ldflags "-s -w $LINKFLAGS" -o bin/vm-dhcp-webhook ./cmd/webhook
+for arch in "amd64" "arm64"; do
+    GOARCH="$arch" CGO_ENABLED=0 GOARCH="$ARCH" go build -ldflags "-s -w $LINKFLAGS" -o bin/vm-dhcp-controller-"$arch" ./cmd/controller
+    GOARCH="$arch" CGO_ENABLED=0 GOARCH="$ARCH" go build -ldflags "-s -w $LINKFLAGS" -o bin/vm-dhcp-agent-"$arch" ./cmd/agent
+    GOARCH="$arch" CGO_ENABLED=0 GOARCH="$ARCH" go build -ldflags "-s -w $LINKFLAGS" -o bin/vm-dhcp-webhook-"$arch" ./cmd/webhook
```
Comment (Yu-Jack) [inline scripts/build]: duplicated

```suggestion
    GOARCH="$arch" CGO_ENABLED=0 go build -ldflags "-s -w $LINKFLAGS" -o bin/vm-dhcp-controller-"$arch" ./cmd/controller
    GOARCH="$arch" CGO_ENABLED=0 go build -ldflags "-s -w $LINKFLAGS" -o bin/vm-dhcp-agent-"$arch" ./cmd/agent
    GOARCH="$arch" CGO_ENABLED=0 go build -ldflags "-s -w $LINKFLAGS" -o bin/vm-dhcp-webhook-"$arch" ./cmd/webhook
```

File: Dockerfile.dapper [hunk 2]
```diff
@@ -6,9 +6,14 @@ ENV ARCH $DAPPER_HOST_ARCH
 RUN zypper -n install tar gzip bash git docker less file curl wget
 
 RUN if [ "${ARCH}" = "amd64" ]; then \
-    curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s v1.55.2; \
+    curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s v1.57.1; \
```
Comment (Yu-Jack) [inline Dockerfile.dapper]: Should we remove `RUN if [ "${ARCH}" = "amd64" ]; then \`?

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+name: Main Build and Publish
+
+on:
+  push:
+    branches:
+    - main
+    - release/v**
```
Comment (bk201) [inline .github/workflows/main.yml]: no `release/v**` branches in this repo.

**Review by Yu-Jack — APPROVED** [review]
LGTM, thanks.

**Comment by starbops** [comment]
The local build path using `make package` seems to be broken with a "file not found" error. We need to add the architecture suffix to the copy command. Please check again, thanks. cc @Yu-Jack

File: scripts/package [hunk 4]
```diff
@@ -9,13 +9,14 @@ echo "Packaging..."
 
 # Copy out artifacts
 mkdir -p dist/artifacts
-cp bin/vm-dhcp-controller dist/artifacts/vm-dhcp-controller${SUFFIX}
+cp bin/vm-dhcp-controller${SUFFIX} dist/artifacts/vm-dhcp-controller${SUFFIX}
```
Comment (Yu-Jack) [inline scripts/package]: you don't need to copy here, Dockerfile will do that for you.

File: scripts/package-agent [hunk 5]
```diff
@@ -9,13 +9,14 @@ echo "Packaging agent..."
 
 # Copy out artifacts
 mkdir -p dist/artifacts
-cp bin/vm-dhcp-agent dist/artifacts/vm-dhcp-agent${SUFFIX}
+cp bin/vm-dhcp-agent${SUFFIX} dist/artifacts/vm-dhcp-agent${SUFFIX}
```
Comment (Yu-Jack) [inline scripts/package-agent]: same

File: scripts/package-webhook [hunk 6]
```diff
@@ -9,13 +9,14 @@ echo "Packaging webhook..."
 
 # Copy out artifacts
 mkdir -p dist/artifacts
-cp bin/vm-dhcp-webhook dist/artifacts/vm-dhcp-webhook${SUFFIX}
+cp bin/vm-dhcp-webhook${SUFFIX} dist/artifacts/vm-dhcp-webhook${SUFFIX}
```
Comment (Yu-Jack) [inline scripts/package-webhook]: same

**Review by Yu-Jack — COMMENTED** [review]
After fixing that, run make build before running make package.

Since Dockerfile wil copy binaries for you, I think we don't need to copy in package.

**Comment by Yu-Jack** [comment]
One more question, do we need artifact folder here?

**Comment by starbops** [comment]
> One more question, do we need artifact folder here?

It seems like a legacy to me. I removed them entirely from the package scripts. I also added a new consolidated `package` script to package all three binaries at once. Thanks for the input!

File: scripts/package [hunk 7]
```diff
@@ -1,30 +1,8 @@
...
-echo Tagged "${IMAGE_PUSH}"
-
-DOCKER_CONFIG=".docker"
-if [[ -n ${PUSH} ]];then
-  docker --config=${DOCKER_CONFIG} push "${IMAGE_PUSH}"
-  echo Pushed "${IMAGE_PUSH}"
-fi
+./package-controller
+./package-agent
+./package-webhook
```
Comment (Yu-Jack) [inline scripts/package:10]: Miss `cd $(dirname $0)`

```suggestion
cd $(dirname $0)
./package-controller
./package-agent
./package-webhook
```

image

**Review by Yu-Jack — CHANGES_REQUESTED** [review]

**Review by bk201 — APPROVED** [review]

**Review by Yu-Jack — APPROVED** [review]
LGTM, thanks!

**Comment by starbops** [comment]
Waiting for rancherlabs/eio#2205. Please do not merge yet.

---

Update: the secrets were migrated.

**Pull request merged**